dart/collision/Result.cpp

```diff
@@ -74,15 +81,73 @@ const std::vector<Contact>& CollisionResult::getContacts() const
 }
 
 //==============================================================================
+const std::unordered_set<const dynamics::BodyNode*>
+CollisionResult::getCollidingBodyNodes() const
+{
+  return mCollidingBodyNodes;
+}
+
+//==============================================================================
+const std::unordered_set<const dynamics::ShapeFrame*>
+CollisionResult::getCollidingShapeFrames() const
+{
+  return mCollidingShapeFrames;
+}
+
+//==============================================================================
+bool CollisionResult::inCollision(const dynamics::BodyNode* bn) const
+{
+  return (mCollidingBodyNodes.find(bn) != mCollidingBodyNodes.end());
+}
+
+//==============================================================================
+bool CollisionResult::inCollision(const dynamics::ShapeFrame* frame) const
+{
+  return (mCollidingShapeFrames.find(frame) != mCollidingShapeFrames.end());
+}
+
+//==============================================================================
 bool CollisionResult::isCollision() const
 {
   return !mContacts.empty();
 }
 
 //==============================================================================
+CollisionResult::operator bool() const
+{
+  return isCollision();
+}
+
+//==============================================================================
 void CollisionResult::clear()
 {
   mContacts.clear();
+  mCollidingShapeFrames.clear();
+  mCollidingBodyNodes.clear();
+}
+
+//==============================================================================
+void CollisionResult::addObject(CollisionObject* object)
+{
+  if(!object)
+  {
+    dterr << "[CollisionResult::addObject] Attempting to add a collision with "
+          << "a nullptr object to a CollisionResult instance. This is not "
+          << "allowed. Please report this as a bug!";
+    assert(false);
+    return;
+  }
+
+  const dynamics::ShapeFrame* frame = object->getShapeFrame();
+  mCollidingShapeFrames.insert(frame);
+
+  if(frame->isShapeNode())
+  {
+    const dynamics::ShapeNode* node =
+        dynamic_cast<const dynamics::ShapeNode*>(frame);
```

Comment:
Yeah, actually I also wondered that when adding `asShapeNode()`. But I found some post like [this](http://stackoverflow.com/a/9959645) saying virtual function is cheaper so added `asShapeNode()` for these use cases.